src/waveform/widgets/allshader/waveformwidget.cpp

```diff
@@ -26,53 +25,70 @@ WaveformWidget::WaveformWidget(QWidget* parent,
 #endif
-    allshader::WaveformRendererSignalBase* waveformSignalRenderer =
-            addWaveformSignalRenderer(
-                    type, options, ::WaveformRendererAbstract::Play);
-
-    addRenderer<WaveformRenderBeat>();
-    addRenderer<WaveformRenderMark>();
-
-    // if the signal renderer supports slip, we add it again, now for slip, together with the
-    // other slip renderers
-    if (waveformSignalRenderer && waveformSignalRenderer->supportsSlip()) {
+    pOpacityNode->appendChildNode(addWaveformSignalRendererNode(
+            type, options, ::WaveformRendererAbstract::Play));
+    pOpacityNode->appendChildNode(addRendererNode<WaveformRenderBeat>());
+    pOpacityNode->appendChildNode(addRendererNode<WaveformRenderMark>());
+    m_pWaveformRenderMark = static_cast<WaveformRenderMark*>(pOpacityNode->lastChild());
+
+    // if the added signal renderer supports slip, we add it again, now for
+    // slip, together with the other slip renderers
+    if (m_pWaveformRendererSignal && m_pWaveformRendererSignal->supportsSlip()) {
         // The following renderer will add an overlay waveform if a slip is in progress
-        addRenderer<WaveformRendererSlipMode>();
-        addRenderer<WaveformRendererPreroll>(::WaveformRendererAbstract::Slip);
+        pOpacityNode->appendChildNode(addRendererNode<WaveformRendererSlipMode>());
+        pOpacityNode->appendChildNode(
+                addRendererNode<WaveformRendererPreroll>(
+                        ::WaveformRendererAbstract::Slip));
 #ifdef __STEM__
-        addRenderer<WaveformRendererStem>(::WaveformRendererAbstract::Slip);
+        pOpacityNode->appendChildNode(
+                addRendererNode<WaveformRendererStem>(
+                        ::WaveformRendererAbstract::Slip));
 #endif
-        addWaveformSignalRenderer(type, options, ::WaveformRendererAbstract::Slip);
-        addRenderer<WaveformRenderBeat>(::WaveformRendererAbstract::Slip);
-        addRenderer<WaveformRenderMark>(::WaveformRendererAbstract::Slip);
+        pOpacityNode->appendChildNode(addWaveformSignalRendererNode(
+                type, options, ::WaveformRendererAbstract::Slip));
+        pOpacityNode->appendChildNode(
+                addRendererNode<WaveformRenderBeat>(
+                        ::WaveformRendererAbstract::Slip));
+        pOpacityNode->appendChildNode(
+                addRendererNode<WaveformRenderMark>(
+                        ::WaveformRendererAbstract::Slip));
     }
 
     m_initSuccess = init();
+
+    pTopNode->appendChildNode(std::move(pOpacityNode));
+    m_pOpacityNode = static_cast<rendergraph::OpacityNode*>(pTopNode->lastChild());
+
+    m_pEngine = std::make_unique<rendergraph::Engine>(std::move(pTopNode));
 }
 
 WaveformWidget::~WaveformWidget() {
     makeCurrentIfNeeded();
-    for (auto* pRenderer : std::as_const(m_rendererStack)) {
-        delete pRenderer;
-    }
     m_rendererStack.clear();
+    m_pEngine.reset();
```

Comment:
is the explicit `.reset` needed for correctness/safety? If so please add a comment explaining why. If not, then the reset should be safe to remove.

Reply:
I don't remember, I will give it a try.

Reply:
the destruction needs to happen inside the OpenGL context.

Reply:
and the context is gone after the end of the user-defined constructor? Or what exactly is the problem?